jupyterhub/app.py

```diff
@@ -2009,22 +2047,35 @@ async def init_roles(self):
                                 )
                         Class = orm.get_class(bearer)
                         orm_obj = Class.find(db, bname)
-                        roles.grant_role(
-                            db, entity=orm_obj, rolename=predef_role['name']
-                        )
-        # make sure that on no admin situation, all roles are reset
-        admin_role = orm.Role.find(db, name='admin')
-        if not admin_role.users:
+                        orm_role_bearers.append(orm_obj)
+                        # Ensure all with admin role have admin flag
+                        if predef_role['name'] == 'admin':
+                            orm_obj.admin = True
+                setattr(predef_role_obj, bearer, orm_role_bearers)
+                db.commit()
```

Comment:
Let's move the commit outside the loop to avoid too many transactions:

```suggestion
        db.commit()
```